examples/linear_regression.rs

```diff
@@ -0,0 +1,126 @@
+#![allow(non_snake_case)]
+use ndarray::{stack, Array, Array1, Array2, ArrayBase, Axis, Data, Ix1, Ix2};
+use ndarray_linalg::{random, Solve};
+use ndarray_rand::RandomExt;
+use ndarray_stats::DeviationExt;
+use rand::distributions::StandardNormal;
+
+/// The simple linear regression model is
+///     y = bX + e  where e ~ N(0, sigma^2 * I)
+/// In probabilistic terms this corresponds to
+///     y - bX ~ N(0, sigma^2 * I)
+///     y | X, b ~ N(bX, sigma^2 * I)
+/// The loss for the model is simply the squared error between the model
+/// predictions and the true values:
+///     Loss = ||y - bX||^2
+/// The maximum likelihood estimation for the model parameters `beta` can be computed
+/// in closed form via the normal equation:
+///     b = (X^T X)^{-1} X^T y
+/// where (X^T X)^{-1} X^T is known as the pseudoinverse or Moore-Penrose inverse.
+///
+/// Adapted from: https://github.com/xinscrs/numpy-ml
+struct LinearRegression {
+    pub beta: Option<Array1<f64>>,
+    fit_intercept: bool,
+}
+
+impl LinearRegression {
+    fn new(fit_intercept: bool) -> LinearRegression {
+        LinearRegression {
+            beta: None,
+            fit_intercept,
+        }
+    }
+
+    fn fit<A, B>(&mut self, X: ArrayBase<A, Ix2>, y: ArrayBase<B, Ix1>)
+    where
+        A: Data<Elem = f64>,
+        B: Data<Elem = f64>,
+    {
+        let (n_samples, _) = X.dim();
+
+        // Check that our inputs have compatible shapes
+        assert_eq!(y.dim(), n_samples);
+
+        // If we are fitting the intercept, we need an additional column
+        self.beta = if self.fit_intercept {
+            let dummy_column: Array<f64, _> = Array::ones((n_samples, 1));
+            let X = stack(Axis(1), &[dummy_column.view(), X.view()]).unwrap();
+            Some(LinearRegression::solve_normal_equation(X, y))
+        } else {
+            Some(LinearRegression::solve_normal_equation(X, y))
+        };
+    }
+
+    fn solve_normal_equation<A, B>(X: ArrayBase<A, Ix2>, y: ArrayBase<B, Ix1>) -> Array1<f64>
```

Comment:
Suggestion: move `solve_normal_equation` out of the public interface of `LinearRegression`

This is a _pure_ function, in that it doesn't twiddle with `self` state directly, so it's safe enough being here. But it does add noise to the interface of the object - the usable surface is `fit` and `predict`.

So as a minor enhancement, can we find somewhere to put both `solve_normal_equation` and `_predict` so they're not exposed?

Reply:
Unfortunately if you do single-file examples everything in a struct is public by default (both methods and fields).
I have broken down the example in two files, with a proper `main`.
Now only what is marked with `pub` can be used when `LinearRegression` is imported in `main`, hence `_predict` and `solve_normal_equation` are private :+1: